src/webgpu/api/operation/storage_texture/read_write.spec.ts

```diff
@@ -0,0 +1,405 @@
+    width: number,
+    height: number,
+    arrayLayers: number
+  ): ArrayBuffer {
+    const bytesPerBlock = kTextureFormatInfo[textureFormat].bytesPerBlock;
+    const initialData = new ArrayBuffer(bytesPerBlock * width * height * arrayLayers);
+
+    for (let z = 0; z < arrayLayers; ++z) {
+      for (let y = 0; y < height; ++y) {
+        for (let x = 0; x < width; ++x) {
+          const index = z * width * height + y * width + x;
+          switch (textureFormat) {
+            case 'r32sint': {
+              const initialSintData = new Int32Array(initialData);
+              initialSintData[index] = 2 * index + 1;
+              break;
+            }
+            case 'r32uint': {
+              const initialUintData = new Uint32Array(initialData);
+              initialUintData[index] = 2 * index + 1;
+              break;
+            }
+            case 'r32float': {
+              const initialFloatData = new Float32Array(initialData);
+              initialFloatData[index] = (2 * index + 1) / 10.0;
+              break;
+            }
+          }
+        }
+      }
+    }
+    return initialData;
+  }
+
+  GetTestSuite(shaderStage: ShaderStageForReadWriteStorageTexture): TestSuite {
+    switch (shaderStage) {
+      case 'compute':
+        return new TestSuiteForCompute();
+      case 'fragment':
+        return new TestSuiteForRendering();
+    }
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('basic')
+  .desc(
+    `The basic functionality tests for read-write storage textures. In the test we read data from
+    the read-write storage texture, write the data back to the read-write storage texture, and
+    finally check the output data by reading the data back to CPU. textureBarrier() is also called
+    in the tests using compute pipelines.`
+  )
+  .params(u =>
+    u
+      .combine('format', kReadWriteStorageTextureFormats)
+      .combine('shaderStage', kShaderStagesForReadWriteStorageTexture)
+      .combine('arrayLayers', [1, 2] as const)
+  )
+  .fn(t => {
+    const { format, shaderStage, arrayLayers } = t.params;
+
+    const kWidth = 16;
+    const kHeight = 8;
+    const kTextureSize = [kWidth, kHeight, arrayLayers] as const;
+    const storageTexture = t.device.createTexture({
+      format,
+      size: kTextureSize,
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST | GPUTextureUsage.STORAGE_BINDING,
+    });
+
+    const bytesPerBlock = kTextureFormatInfo[format].bytesPerBlock;
+    const initialData = t.GetInitialData(format, kWidth, kHeight, arrayLayers);
+    t.queue.writeTexture(
+      { texture: storageTexture },
+      initialData,
+      {
+        bytesPerRow: bytesPerBlock * kWidth,
+        rowsPerImage: kHeight,
+      },
+      kTextureSize
+    );
+
+    const commandEncoder = t.device.createCommandEncoder();
+
+    const testSuite = t.GetTestSuite(shaderStage);
+    testSuite.RecordCommandsToTransform(t.device, commandEncoder, storageTexture, format);
```

Comment:
nit: Maybe we could have compute | render, swithc on it, and have a single function that both encodes on the encoder, and returns the expected data? It would avoid having the "TestSuite" concept that seems to overload with TestGroup.